platform-sdk/platform-apps/tests/ConsistencyTestingTool/src/main/java/com/swirlds/demo/consistency/TransactionHandlingHistory.java

```diff
@@ -0,0 +1,290 @@
+     */
+    @Nullable
+    private String compareWithHistoricalRound(
+            final @NonNull ConsistencyTestingToolRound newRound,
+            final @NonNull ConsistencyTestingToolRound historicalRound) {
+
+        Objects.requireNonNull(newRound);
+        Objects.requireNonNull(historicalRound);
+
+        if (!newRound.equals(historicalRound)) {
+            final String error =
+                    "Round " + newRound.roundNumber() + " with transactions " + newRound.transactionsContents()
+                            + " doesn't match historical counterpart with transactions "
+                            + historicalRound.transactionsContents();
+
+            logger.error(EXCEPTION.getMarker(), error);
+            return error;
+        }
+
+        return null;
+    }
+
+    /**
+     * Add a round to the history. Errors are logged.
+     *
+     * @param newRound the round to add
+     * @return a list of errors that occurred while adding the round
+     */
+    @NonNull
+    private List<String> addRoundToHistory(final @NonNull ConsistencyTestingToolRound newRound) {
+        Objects.requireNonNull(newRound);
+
+        final List<String> errors = new ArrayList<>();
+
+        roundHistory.add(newRound);
+
+        newRound.transactionsContents().forEach(transaction -> {
+            final String error = addTransaction(transaction);
+
+            if (error != null) {
+                errors.add(error);
+            }
+        });
+
+        if (roundHistory.size() <= 1) {
+            // only 1 round is in the history, so no additional checks are necessary
+            return errors;
+        }
+
+        final long newRoundNumber = newRound.roundNumber();
+        final long previousRoundNumber =
+                roundHistory.get(roundHistory.size() - 2).roundNumber();
+
+        // make sure round numbers always increase
+        if (newRoundNumber <= previousRoundNumber) {
+            final String error = "Round " + newRoundNumber + " is not greater than round " + previousRoundNumber;
+            logger.error(EXCEPTION.getMarker(), error);
+
+            errors.add(error);
+        }
+        // if gaps in round history aren't permitted, check that the round numbers are consecutive
+        else if (!permitRoundGaps && newRoundNumber != previousRoundNumber + 1) {
+            final String error = "Gap in round history found. Round " + newRoundNumber
+                    + " was added, but previous round was " + previousRoundNumber;
+
+            logger.error(EXCEPTION.getMarker(), error);
+
+            errors.add(error);
+        }
+
+        return errors;
+    }
+
+    /**
+     * Writes the given round to the log file
+     *
+     * @param round the round to write to the log file
+     */
+    private void writeRoundToLog(final @NonNull ConsistencyTestingToolRound round) {
+        Objects.requireNonNull(round);
+
+        try (BufferedWriter file = new BufferedWriter(new FileWriter(logFilePath.toFile(), true))) {
```

Comment:
It would be more performant if we didn't open and close the file each round (although the flush is still necessary).